TYPO3.Neos/Classes/TYPO3/Neos/Utility/NodeUriPathSegmentGenerator.php

```diff
@@ -31,22 +33,44 @@ class NodeUriPathSegmentGenerator
     protected $transliterationService;
 
     /**
+     * @Flow\Inject
+     * @var NodeTypeManager
+     */
+    protected $nodeTypeManager;
+
+    /**
+     * @Flow\Inject
+     * @var NodeSearchServiceInterface
+     */
+    protected $nodeSearchService;
+
+    /**
      * Sets the best possible uriPathSegment for the given Node.
      * Will use an already set uriPathSegment or alternatively the node name as base,
      * then checks if the uriPathSegment already exists on the same level and appends a counter until a unique path segment was found.
      *
      * @param NodeInterface $node
      * @return void
      */
-    public static function setUniqueUriPathSegment(NodeInterface $node)
+    public function setUniqueUriPathSegment(NodeInterface $node)
     {
         if ($node->getNodeType()->isOfType('TYPO3.Neos:Document')) {
-            $q = new FlowQuery(array($node));
-            $q = $q->context(array('invisibleContentShown' => true, 'removedContentShown' => true, 'inaccessibleContentShown' => true));
-
+            $parentNode = $node->getParent();
+            $documentNodeType = $this->nodeTypeManager->getNodeType('TYPO3.Neos:Document');
+            $context = $node->getContext();
             $possibleUriPathSegment = $initialUriPathSegment = !$node->hasProperty('uriPathSegment') ? $node->getName() : $node->getProperty('uriPathSegment');
+
+            $nodeExists = function ($possibleUriPathSegment) use ($documentNodeType, $context, $parentNode, $node) {
+                $nodes = $this->nodeSearchService->findByProperties(['uriPathSegment' => $possibleUriPathSegment], [$documentNodeType], $context, $parentNode);
```

Comment:
@aertmann I'll definitely do tests (maybe today). What do you think about adding an option to the search service to only search for direct children? That would be one LIKE condition less in the query, and will also simplify the results handling (no need to filter for direct children).